db/migrate/20220118053107_convert_stripe_connect_to_stripe_sca.rb

```diff
@@ -0,0 +1,57 @@
+class ConvertStripeConnectToStripeSca < ActiveRecord::Migration[6.1]
+  class SpreePreference < ActiveRecord::Base
+    scope :leftover_from_payment_type, ->(class_name) {
+      joins <<~SQL
+         JOIN spree_payment_methods
+           ON spree_preferences.key =
+              CONCAT('/#{class_name.underscore}/enterprise_id/', spree_payment_methods.id)
+          AND spree_payment_methods.type != '#{class_name}'
+      SQL
+    }
+  end
+
+  def up
+    delete_outdated_spree_preferences
+    upgrade_stripe_payment_methods
+    update_payment_method_preferences
+  end
+
+  private
+
+  # When changing the type of a payment method, we leave orphaned records in
+  # the spree_preferences table. The key of the preference contains the type
+  # of the payment method and therefore changing the type disconnects the
+  # preference.
```

Comment:
That sounds like it could cause nasty issues. Is this orphaned-preferences thing something that needs cleaning up (in general, not in this PR)?

Reply:
This is the way the Admin UX works. It makes sure though to always override stale values when needed. So we don't think that we have nasty issues at the moment. But in general, it's not good to change the type of a polymorphic model with spree preferences. The better way would be to not allow that and create a new payment method even though that would have made the current Stripe migration more difficult.

So basically, let's phase out Spree Preferences, I'd say. Not in this PR. :smile:

Reply:
Just for reference:
https://github.com/openfoodfoundation/openfoodnetwork/blob/b86e2c8c16252679f743bc19d02018c013f9e406/app/controllers/spree/admin/payment_methods_controller.rb#L34-L47

The type change is done first which leaves orphaned data. Then we load the new class and update with all the params which should override any stale data that would be conflicting (if you changed the type and then back again). So the only problem we have is orphaned data in the table.